numpy/_core/src/umath/ufunc_object.c

```diff
@@ -6411,6 +6412,19 @@ ufunc_get_doc(PyUFuncObject *ufunc, void *NPY_UNUSED(ignored))
 {
     PyObject *doc;
 
+    // if there is a __doc__ in the instance __dict__, use that, but only if
+    // __doc__ is NULL
+    if (ufunc->doc == NULL) {
```

Comment:
I think I might reverse the logic: if `__doc__` is present in `ufunc->dict` return that. If not, use `ufunc->doc` as now to calculate the final docstring, and then store the result in  `__doc__` (so that it gets re-used any further times, unless it is overwritten; and if one does `del ufunc.__doc__`, one will automatically get the original docstring back).